internal/provider/resource_storagezone.go

```diff
@@ -0,0 +1,397 @@
+	clt := meta.(*bunny.Client)
+
+	id, err := getIDAsInt64(d)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	sz, err := clt.StorageZone.Get(ctx, id)
+	if err != nil {
+		return diagsErrFromErr("could not retrieve storage zone", err)
+	}
+
+	if err := storageZoneToResource(sz, d); err != nil {
+		return diagsErrFromErr("converting api type to resource data after successful read failed", err)
+	}
+
+	return nil
+}
+
+func resourceStorageZoneDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	clt := meta.(*bunny.Client)
+
+	id, err := getIDAsInt64(d)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	err = clt.StorageZone.Delete(ctx, id)
+	if err != nil {
+		return diagsErrFromErr("could not delete storage zone", err)
+	}
+
+	d.SetId("")
+
+	return nil
+}
+
+// storageZoneToResource sets fields in d to the values in sz.
+func storageZoneToResource(sz *bunny.StorageZone, d *schema.ResourceData) error {
+	if sz.ID != nil {
+		d.SetId(strconv.FormatInt(*sz.ID, 10))
+	}
+
+	if err := d.Set(keyUserID, sz.UserID); err != nil {
+		return err
+	}
+	if err := d.Set(keyName, sz.Name); err != nil {
+		return err
+	}
+	if err := d.Set(keyPassword, sz.Password); err != nil {
+		return err
+	}
+	if err := d.Set(keyDateModified, sz.DateModified); err != nil {
+		return err
+	}
+	if err := d.Set(keyDeleted, sz.Deleted); err != nil {
+		return err
+	}
+	if err := d.Set(keyStorageUsed, sz.StorageUsed); err != nil {
+		return err
+	}
+	if err := d.Set(keyFilesStored, sz.FilesStored); err != nil {
+		return err
+	}
+	if err := d.Set(keyRegion, sz.Region); err != nil {
+		return err
+	}
+	if err := d.Set(keyReadOnlyPassword, sz.ReadOnlyPassword); err != nil {
+		return err
+	}
+	if err := setStrSet(d, keyReplicationRegions, sz.ReplicationRegions, ignoreOrderOpt, caseInsensitiveOpt); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func revertUpdateValues(d *schema.ResourceData) error {
+	o, _ := d.GetChange(keyOriginURL)
+	if err := d.Set(keyOriginURL, o); err != nil {
+		return err
+	}
+	o, _ = d.GetChange(keyCustom404FilePath)
+	if err := d.Set(keyCustom404FilePath, o); err != nil {
+		return err
+	}
+	o, _ = d.GetChange(keyRewrite404To200)
+	if err := d.Set(keyRewrite404To200, o); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// storageZoneFromResource returns a StorageZoneUpdateOptions API type that
+// has fields set to the values in d.
+func storageZoneFromResource(d *schema.ResourceData) (*bunny.StorageZoneUpdateOptions, error) {
+	var res bunny.StorageZoneUpdateOptions
+
+	res.ReplicationRegions = getStrSetAsSlice(d, keyReplicationRegions)
+
+	if d.HasChange(keyOriginURL) {
+		res.OriginURL = getStrPtr(d, keyOriginURL)
+	}
+
+	if d.HasChange(keyCustom404FilePath) {
+		res.Custom404FilePath = getStrPtr(d, keyCustom404FilePath)
+	}
+
+	if d.HasChange(keyRewrite404To200) {
+		res.Rewrite404To200 = getBoolPtr(d, keyRewrite404To200)
+	}
+
+	return &res, nil
```

Comment:
I did it this way due to some finicky behavior where the values were getting defaulted to an empty string if empty rather than `nil` which causes a bunny error. So as long as that still happens I'm all for it.

Reply:
I fixed those issues by using getOk instead get when retrieving the values from the resource (https://github.com/5-stones/terraform-provider-bunny/pull/1/commits/773b9beca8d5ae51bf7b3279f019e6dceedcb96c).